catalog/views.py

```diff
@@ -170,3 +198,66 @@ class BookDelete(PermissionRequiredMixin, DeleteView):
     model = Book
     success_url = reverse_lazy('books')
     permission_required = 'catalog.can_mark_returned'
+
+    def form_valid(self, form):
+        try:
+            self.object.delete()
+            return HttpResponseRedirect(self.success_url)
+        except Exception as e:
+            return HttpResponseRedirect(
+                reverse("book-delete", kwargs={"pk": self.object.pk})
+            )
```

Comment:
This is the code that prevents the error if you try to delete something and an exception is raised - usually if it is protected. 
Normally perhaps you'd feed in an error and re-render but there is no "official" way to do this with a model view. I might look at a flash message, or some other approach.
What I think I will do is make the template render itself stating what has to be deleted (always), so this just puts you back to the same place. Not implemented yet.